Filter rules gui

CautemocSg opened the pull request:
Adds an in-extension user interface for editing and adding filter rules previously handled by the JSON editor

CautemocSg committed:
Initial commit

CautemocSg committed:
Revert "Initial commit"

This reverts commit 6c4ccb1c4403d4cabb0a4518aaf522368e84619f.

CautemocSg committed:
Add filter rules GUI

CautemocSg committed:
Merge branch 'main' into filter-rules-gui

CautemocSg committed:
Package update

CautemocSg committed:
Update README.md

CautemocSg committed:
Refactored details sections

CautemocSg committed:
Update ToolbarButton.tsx

CautemocSg committed:
Merge branch 'main' into filter-rules-gui

errnoh commented:
Doesn't seem to open in Firefox (probably good idea to check other browsers as well, for example there are bunch of users using some android browsers)

danreeves commented on src/components/RuleBasedFilters/RuleBasedFilters.tsx:
```diff
@@ -0,0 +1,189 @@
+                  value={props.DE}
+                  onChange={(event) => {
+                    props.setDE(event.target.value as DeemphasizeOption)
+                    localStorage.setItem("deemphasize-selection", event.target.value)
```
Local storage access was moved to a `useLocalStorage` hook which namespaces our keys

CautemocSg replied:
Fixed, re-implemented as useLocalStorage

danreeves commented on src/types.ts:
```diff
@@ -223,5 +226,21 @@ export interface FilterRule {
+  isOpen: boolean
+}
+
+export enum Size {
+  Small,
+  Medium,
+  Large
+}
```
Not a fan of enums, this could be a string union

CautemocSg replied:
Fixed, re-implemented as string union

danreeves reviewed (commented):
I'm not a fan of adding such a giant client side dependency as styled-components. What do you think about using something like linaria where we can generate the css at build time https://github.com/callstack/linaria

CautemocSg committed:
Removed styled-components and major refactor work

CautemocSg committed:
Updated de-emphasis selector to useLocalStorage

CautemocSg committed:
Refactored Size type to string union

CautemocSg commented:
> Doesn't seem to open in Firefox (probably good idea to check other browsers as well, for example there are bunch of users using some android browsers)

It was broken in Firefox, latest should be working now. Tested on my desktop is working as intended. Could use a check on phone browsers.

errnoh commented:
Yup, seems to be working on Firefox now.

I initially had some problems understanding the save button is at top. If you have bunch of rules you'll need to scroll upwards to find the save button. Should it be at the bottom?

positive:
+ I like that it updates JSON instantly when saving
+ UI themed with the site :)

negative:
- JSON is a bit verbose with all fields being visible with default zero values 
- Not yet on feature parity with the current available rbf rules
- The rule string doesn't seem to show everything, I have some rules that show as empty even when they have some rules defined. (e.g. rule with only blessing rarity defined)

errnoh commented:
And a feature wish: allow dragging rules to reorder them. Wasn't relevant previously, but with the `color` option it's a proper QoL improvement :)

CautemocSg committed:
Added new filters from recent updates

CautemocSg committed:
Merge branch 'main' into filter-rules-gui

CautemocSg commented:
Addressed these points with recent update:

> * JSON is a bit verbose with all fields being visible with default zero values
> * Not yet on feature parity with the current available rbf rules
> * The rule string doesn't seem to show everything, I have some rules that show as empty even when they have some rules defined. (e.g. rule with only blessing rarity defined)

errnoh commented:
Great changes! Hopefully I wasn't too blunt with my feedback :) Just wanting to be sure that such a major UI overhaul feels intuitive to the users.

errnoh commented:
Couple additional things that might be good features. Maybe not worth scoping in on this PR, but related:
- Have the highlight color visible in the rules listing somehow (not just edit window). Small box with that color? Text on the highlight color? I've been locally using the coloring rules for some time already and while it has been one of the largest QoL improvements I still sometimes forget what each color actually means, I'd assume others will run into the same issue.
- Autocomplete / validation. For some fields it's quite simple as the possible values are limited by the type, for some we could do suggestions based on data available (blessing/perk names etc). Then again for weapons there are a lot of cases where one might not want to use full item names, while for blessings that's actually usually the case.

CautemocSg committed:
Added drag and drop functionality for reordering rules and color icon prefix for headers

CautemocSg committed:
Merge branch 'filter-rules-gui' of https://github.com/CautemocSg/dt-exchange into filter-rules-gui

CautemocSg commented:
>And a feature wish: allow dragging rules to reorder them. Wasn't relevant previously, but with the color option it's a proper QoL improvement :)

>Have the highlight color visible in the rules listing somehow (not just edit window). Small box with that color?

@errnoh  Thanks for the feedback and suggestions.

I implemented these 2 requests in the latest update.

CautemocSg committed:
Update RuleBasedFilters.tsx

errnoh commented:
Replied on Discord but I'll add it to discussion here as well:

👍 from me functionality-wise. It's in a state that I have no need to manually edit JSON other than to test things and share configs between browsers & machines. I think if we can get someone to thumbs up the code changes this should be ok to merge?

Btw I've tested this on Firefox and Chrome. If someone has a setup where they can test that it works fine with the android browser bunch of people are using that would be a good thing to check.

danreeves reviewed (approved):
This is awesome!
